src/validators.py

```diff
@@ -0,0 +1,82 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import logging
+from functools import wraps
+from typing import Any, Callable
+
+from ops.charm import CharmBase, EventBase
+from ops.model import BlockedStatus, MaintenanceStatus, WaitingStatus
+
+logger = logging.getLogger(__name__)
+
+
+def leader_unit(func: Callable):
+    @wraps(func)
+    def wrapper(self: CharmBase, *args: EventBase, **kwargs: Any):
+        if not self.unit.is_leader():
+            return
+
+        return func(self, *args, **kwargs)
+
+    return wrapper
+
+
+def validate_container_connectivity(func: Callable):
+    @wraps(func)
+    def wrapper(self: CharmBase, *args: EventBase, **kwargs: Any):
+        event, *_ = args
+        logger.debug(f"Handling event: {event}")
+        if not self._container.can_connect():
+            logger.debug(f"Cannot connect to container, defer event {event}.")
+            event.defer()
+
+            self.unit.status = WaitingStatus("Waiting to connect to container.")
+            return
+
+        return func(self, *args, **kwargs)
+
+    return wrapper
+
+
+def validate_integration_exists(integration_name: str):
+    def decorator(func: Callable):
+        @wraps(func)
+        def wrapper(self: CharmBase, *args: EventBase, **kwargs: Any):
+            event, *_ = args
+            logger.debug(f"Handling event: {event}")
+
+            self.unit.status = MaintenanceStatus("Configuring resources")
```

Comment:
Correct me if I'm wrong, but you don't know if the wrapped function will actually configure resources. Perhaps the inner func should set this status if needed.